x/cron/cron.go

```diff
@@ -0,0 +1,273 @@
+type Ticker struct {
+	hn      weave.Handler
+	enc     TaskMarshaler
+	results orm.ModelBucket
+}
+
+var _ weave.Ticker = (*Ticker)(nil)
+
+// Tick implementes weave.Ticker interface.
+//
+// Tick can process any number of messages suitable for execution. All changes
+// are done atomically and apply only on success.
+func (t *Ticker) Tick(ctx context.Context, db store.CacheableKVStore) [][]byte {
+	executed, err := t.tick(ctx, db)
+	if err != nil {
+		// This is a hopeless state. This error is most likely due to a
+		// database issues or some other instance specific problems.
+		// This problem is unique to this instance and this operation
+		// most likely succeeded on other nodes. This means that there
+		// is no way we could continue operating as this instance is
+		// out of sync with the rest of the network.
+		failTask(err)
+	}
+	return executed
+}
+
+// failTask is a variable so that it can be overwritten for tests.
+var failTask = func(err error) {
+	panic(fmt.Sprintf(`
+
+Asynchronous task failed.
+
+This error is most likely due to a database issues or some other instance
+specific problems. This problem is unique to this instance and this operation
+most likely succeeded on other nodes. This means that there is no way we could
+continue operating as this instance is out of sync with the rest of the
+network.
+
+%+v
+
+	`, err))
+}
+
+// tick process any number of tasks. It always returns a response and might
+// return an error. This method is similar to the Tick except it provides an
+// error. This makes it easier for the tests to check the result.
+func (t *Ticker) tick(ctx context.Context, db store.CacheableKVStore) ([][]byte, error) {
+	var executed [][]byte
+
+	now, err := weave.BlockTime(ctx)
+	if err != nil {
+		return executed, errors.Wrap(err, "cannot get current time")
+	}
+
+	for {
+		switch key, raw, err := peek(db, now); {
+		case err == nil:
+			res := TaskResult{
+				Metadata:   &weave.Metadata{Schema: 1},
+				Successful: true,
+			}
+
+			// Each task is processed using its own cache instance
+			// to ensure changes are atomic and task processing
+			// independent.
+			cache := db.CacheWrap()
+
+			auth, msg, err := t.enc.UnmarshalTask(raw)
+			if err != nil {
+				res.Successful = false
+				res.Info = fmt.Sprintf("cannot unmarshal task: %+v", err)
+			} else {
+				taskCtx := withAuth(ctx, auth)
+				tx := &taskTx{msg: msg}
+				if _, err := t.hn.Deliver(taskCtx, cache, tx); err != nil {
+					// Discard any changes that the deliver could
+					// have created. We do not want to persist
+					// those.
+					cache.Discard()
+					res.Successful = false
+					res.Info = err.Error()
+				}
+			}
+
+			if _, err := t.results.Put(cache, key, &res); err != nil {
+				// Keep it atomic.
+				cache.Discard()
+				return executed, errors.Wrap(err, "cannot store result")
+			}
+
+			// Remove the task from the queue as it was processed.
+			// Do it via cache to keep it atomic.
+			cache.Delete(key)
```

Comment:
This never returns error? Or you ignored that result?

Reply:
Good catch! It does return error and I should handle it.